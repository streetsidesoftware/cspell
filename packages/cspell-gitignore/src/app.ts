import { GitIgnore, findRepoRoot } from '.';

import * as path from 'path';

type OptionParser = (params: string[]) => string[];

const helpText = `Usage cspell-gitignore [options] <files>

Check files against .gitignore
Compare against git check-ignore -v -n <files>

Options:
-r, --root   Add a root to prevent searching for .gitignore files above the root if the file is under the root.
             This option can be used multiple times to add multiple roots. The default root is the current
             repository root determined by the \`.git\` directory.

Example:
  cspell-gitignore README.md
  cspell-gitignore -r . node_modules
`;

export async function run(args: string[]): Promise<void> {
    const { roots, files, help } = parseArgs(args.slice(2));
    const cwd = process.cwd();
    const repo = (await findRepoRoot(cwd)) || cwd;
    const gi = await createGitIgnore(roots, repo);

    if (help) {
        console.log(help);
        return;
    }

    if (!files.length) {
        console.error('Missing files');
        process.exitCode = 1;
        return;
    }

    for (const file of files) {
        const filename = path.relative(cwd, file);
        const pFile = gi.isIgnoredEx(file);
        const pDir = gi.isIgnoredEx(file + '/ ');
        const r = (await pFile) || (await pDir);
        const gitignore = r?.gitIgnoreFile ? path.relative(repo, r.gitIgnoreFile) : '';
        const line = r?.line || '';
        const glob = r?.glob || '';
        console.log(`${gitignore}:${line}:${glob}\t${filename}`);
    }
}

function parseArgs(params: string[]) {
    const roots: string[] = [];
    const files: string[] = [];
    let help = '';

    const options: Record<string, OptionParser | undefined> = {
        '-r': optionRoot,
        '--root': optionRoot,
        '-h': optionHelp,
        '--help': optionHelp,
    };

    function optionRoot(params: string[]): string[] {
        const root = params[1];
        if (!root) {
            throw new Error('Missing root parameter.');
        }
        roots.push(path.resolve(root));
        return params.slice(2);
    }

    function optionFile(params: string[]): string[] {
        const file = params[0].trim();

        if (file) {
            files.push(path.resolve(file));
        }

        return params.slice(1);
    }

    function optionHelp(_params: string[]) {
        help = helpText;
        return [];
    }

    while (params.length) {
        const fn = options[params[0]];
        params = fn?.(params) ?? optionFile(params);
    }

    return { roots, files, help };
}

async function createGitIgnore(roots: string[], repoRoot: string) {
    if (!roots.length) {
        roots.push(repoRoot);
    }

    return new GitIgnore(roots);
}
