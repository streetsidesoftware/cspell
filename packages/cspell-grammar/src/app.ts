import type { ParsedText, Parser } from '@cspell/cspell-types/Parser';
import { promises as fs } from 'fs';
import * as path from 'path';
import { parser as parserTypeScript } from './parsers/typescript';

const parsers: Record<string, Parser> = {
    '.ts': parserTypeScript,
};

/**
 * Run simple parser
 * @param args -- command line arguments
 * @returns Promise
 */
export async function run(args: string[]): Promise<void> {
    // early out if there are not enough arguments
    if (args.length < 3) {
        console.log('usage...');
        return;
    }

    const filename = args.slice(2).filter((p) => !p.startsWith('-'))[0];
    if (!filename) {
        console.log('filename missing');
        return;
    }

    const ext = path.extname(filename);
    const parser = parsers[ext];
    if (!parser) {
        console.log(`No parser for ${path.basename(filename)}`);
        return;
    }

    console.log(`File: ${path.basename(filename)} Parser: ${parser.name}`);
    const content = await fs.readFile(filename, 'utf-8');

    const result = parser.parse(content, filename);
    for (const pt of result.parsedTexts) {
        emit(pt);
    }
}

function emit(pt: ParsedText) {
    const t = pt.text.replace(/\t/g, '↦').replace(/\r?\n/g, '↩︎').replace(/\r/g, '⇠');
    console.log(`${pt.range[0]}-${pt.range[1]}\t${t}\t${pt.scope?.toString() || ''}`);
}
