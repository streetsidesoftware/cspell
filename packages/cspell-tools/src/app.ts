#!/usr/bin/env node

// For large dictionaries, it is necessary to increase the memory limit.
// # !/usr/bin/env node --max_old_space_size=8192


import { compileWordList, compileTrie } from './compiler';
import * as path from 'path';
import * as program from 'commander';
import * as glob from 'glob';
import { genSequence, Sequence } from 'gensequence';
import { streamWordsFromFile } from './compiler/iterateWordsFromFile';
const npmPackage = require(path.join(__dirname, '..', 'package.json'));

function globP(pattern: string): Promise<string[]> {
    return new Promise((resolve, reject) => {
        glob(pattern, (err, result) => {
            err ? reject(err) : resolve(result);
        });
    });
}

interface CompileCommonOptions {
    output?: string;
    compress: boolean;
    case: boolean;
    max_depth?: string;
    merge: string;
}

interface CompileOptions extends CompileCommonOptions {
    split: boolean;
    sort: boolean;
}

interface CompileTrieOptions extends CompileCommonOptions {
    trie3: boolean;
}

export function run(
    program: program.Command,
    argv: string[]
): Promise<void> {
    (program as any).exitOverride();

    return new Promise((resolve, rejects) => {
        program
            .version(npmPackage.version);

        program
            .command('compile <src...>')
            .description('compile words lists into simple dictionary files.')
            .option('-o, --output <path>', 'Specify the output directory, otherwise files are written back to the same location.')
            .option('-n, --no-compress', 'By default the files are Gzipped, this will turn that off.')
            .option('-m, --max_depth <limit>', 'Maximum depth to apply suffix rules.')
            .option('-M, --merge <target>', 'Merge all files into a single target file (extensions are applied)')
            .option('-s, --no-split', 'Treat each line as a dictionary entry, do not split')
            .option('--no-sort', 'Do not sort the result')
            .action((src: string[], options: CompileOptions) => {
                const result = processAction(src, '.txt', options, async (src, dst) => {
                    return compileWordList(src, dst, { splitWords: options.split, sort: options.sort }).then(() => src);
                });
                resolve(result);
            });

        program
            .command('compile-trie <src...>')
            .description('Compile words lists or Hunspell dictionary into trie files used by cspell.')
            .option('-o, --output <path>', 'Specify the output directory, otherwise files are written back to the same location.')
            .option('-m, --max_depth <limit>', 'Maximum depth to apply suffix rules.')
            .option('-M, --merge <target>', 'Merge all files into a single target file (extensions are applied)')
            .option('-n, --no-compress', 'By default the files are Gzipped, this will turn that off.')
            .option('--trie3', '[Beta] Use file format trie3')
            .action((src: string[], options: CompileTrieOptions) => {
                const result = processAction(src, '.trie', options, async (words: Sequence<string>, dst) => {
                    return compileTrie(words, dst, options);
                });
                resolve(result);
            });

        try {
            program.parse(argv);
            if (!argv.slice(2).length) {
                program.help();
            }
        } catch (e) {
            rejects(e);
        }

        resolve();
    });
}

interface FileToProcess {
    src: string;
    words: Sequence<string>;
}

async function processAction(
    src: string[],
    fileExt: '.txt' | '.trie',
    options: CompileCommonOptions,
    action: (words: Sequence<string>, dst: string) => Promise<any>)
: Promise<void> {
    console.log('Compile:\n output: %s\n compress: %s\n files:\n  %s \n\n',
        options.output || 'default',
        options.compress ? 'true' : 'false',
        src.join('\n  '));

    const ext = fileExt + (options.compress ? '.gz' : '');
    const { max_depth } = options;
    const maxDepth = max_depth !== undefined ? Number.parseInt(max_depth) : undefined;
    const readerOptions = { maxDepth };

    const globResults = await Promise.all(src.map(s => globP(s)));
    const filesToProcess = genSequence(globResults)
        .concatMap(files => files)
        .map(async s => {
            const words = await streamWordsFromFile(s, readerOptions);
            const f: FileToProcess = {
                src: s,
                words,
            };
            return f;
        });

    const r = options.merge
    ? processFiles(action, filesToProcess, toMergeTargetFile(options.merge, options.output, ext))
    : processFilesIndividually(action, filesToProcess, s => toTargetFile(s, options.output, ext));
    await r;
    console.log(`Complete.`);
}

function toFilename(name: string, ext: string) {
    return path.basename(name).replace(/((\.txt|\.dic|\.aff|\.trie)(\.gz)?)?$/, '') + ext;
}

function toTargetFile(filename: string, destination: string | undefined, ext: string) {
    const outFileName = toFilename(filename, ext);
    const dir = destination ?? path.dirname(filename);
    return path.join(dir, outFileName);
}

function toMergeTargetFile(filename: string, destination: string | undefined, ext: string) {
    const outFileName = toFilename(filename, ext);
    return path.resolve(destination ?? './', outFileName);
}

async function processFilesIndividually(
    action: (words: Sequence<string>, dst: string) => Promise<any>,
    filesToProcess: Sequence<Promise<FileToProcess>>,
    srcToTarget: (src: string) => string,
) {
    const toProcess = filesToProcess
    .map(async pFtp => {
        const { src, words } = await pFtp;
        const dst = srcToTarget(src);
        console.log('Process "%s" to "%s"', src, dst);
        await action(words, dst);
        console.log('Done "%s" to "%s"', src, dst);
    });

    for (const p of toProcess) {
        await p;
    }
}

async function processFiles(
    action: (words: Sequence<string>, dst: string) => Promise<any>,
    filesToProcess: Sequence<Promise<FileToProcess>>,
    mergeTarget: string,
) {
    const toProcess = await Promise.all([...filesToProcess]);
    const dst = mergeTarget;

    const words = genSequence(toProcess)
    .map(ftp => {
        const { src } = ftp;
        console.log('Process "%s" to "%s"', src, dst);
        return ftp;
    })
    .concatMap( ftp => ftp.words );
    await action(words, dst);
    console.log('Done "%s"', dst);
}

if (require.main === module) {
    run(program, process.argv).catch(() => process.exit(1));
}
