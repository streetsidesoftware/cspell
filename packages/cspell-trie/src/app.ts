import type * as commander from 'commander';
import * as Trie from 'cspell-trie-lib';
import { createWriteStream as fsCreateWriteStream } from 'fs';
import * as fsp from 'fs/promises';
import type { Sequence } from 'gensequence';
import { genSequence } from 'gensequence';
import * as path from 'path';
import * as stream from 'stream';
import { fileURLToPath } from 'url';
import * as zlib from 'zlib';

const UTF8: BufferEncoding = 'utf8';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

interface PackageJson {
    version: string;
}

async function getPackageInfo(): Promise<PackageJson> {
    const packageInfo = await fsp.readFile(path.join(__dirname, '../package.json'), 'utf8');
    return JSON.parse(packageInfo);
}

async function getPackageVersion() {
    return (await getPackageInfo()).version;
}

export async function run(program: commander.Command, argv: string[]): Promise<commander.Command> {
    const version = await getPackageVersion();
    program.version(version);

    program
        .command('create <file.txt>')
        .option('-o, --output <file>', 'output file - defaults to stdout')
        .option('-l, --lower_case', 'output in lower case')
        .option(
            '-b, --base <number>',
            'Use base n for reference ids.  Defaults to 32. Common values are 10, 16, 32. Max of 36'
        )
        .description('Generate a file for use with cspell')
        .action(async (filename, options) => {
            const { output: outputFile, lower_case: lowerCase = false, base = 32 } = options;
            notify('Create Trie', !!outputFile);
            const pOutputStream = createWriteStream(outputFile);
            notify(`Generating...`, !!outputFile);
            const lines = await fileToLines(filename);
            const toLower = lowerCase ? (a: string) => a.toLowerCase() : (a: string) => a;

            const wordsRx = lines
                .map(toLower)
                .map((a) => a.trim())
                .filter((a) => !!a);

            notify('Processing Trie');
            const trie = Trie.buildTrie(wordsRx);

            notify('Export Trie');
            const serialStream = stream.Readable.from(Trie.serializeTrie(trie.root, base - 0 || 32));
            const outputStream = await pOutputStream;
            return new Promise((resolve) => {
                serialStream.pipe(outputStream).on('finish', () => resolve());
            });
        });

    program
        .command('reader <file.trie>')
        .option('-o, --output <file>', 'output file - defaults to stdout')
        .description('Read a cspell trie file and output the list of words.')
        .action(async (filename, options) => {
            const { output: outputFile } = options;
            notify('Reading Trie', !!outputFile);
            const pOutputStream = createWriteStream(outputFile);
            const lines = await fileToLines(filename);
            const root = Trie.importTrie(lines);
            const words: Sequence<string> = Trie.iteratorTrieWords(root);
            const outputStream = await pOutputStream;
            return new Promise((resolve) => {
                stream.Readable.from(words.map((a) => a + '\n'))
                    .pipe(outputStream)
                    .on('finish', () => resolve());
            });
        });

    try {
        return program.parseAsync(argv);
    } catch (e) {
        return Promise.reject(e);
    }
}

async function fileToLines(filename: string): Promise<Sequence<string>> {
    const buffer = await fsp.readFile(filename);
    const file = (filename.match(/\.gz$/) ? zlib.gunzipSync(buffer) : buffer).toString(UTF8);
    return genSequence(file.split(/\r?\n/));
}

function createWriteStream(filename?: string): Promise<NodeJS.WritableStream> {
    return !filename
        ? Promise.resolve(process.stdout)
        : fsp.mkdir(path.dirname(filename), { recursive: true }).then(() => fsCreateWriteStream(filename));
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function notify(message: any, useStdOut = true) {
    if (useStdOut) {
        console.log(message);
    } else {
        console.error(message);
    }
}
