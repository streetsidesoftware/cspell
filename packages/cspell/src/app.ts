#!/usr/bin/env node

import * as path from 'path';
import * as commander from 'commander';
const npmPackage = require(path.join(__dirname, '..', 'package.json'));
import { CSpellApplicationOptions, BaseOptions, checkText } from './application';
import * as App from './application';
import chalk = require('chalk');

interface Options extends CSpellApplicationOptions {
    legacy?: boolean;
    summary: boolean;
    issues: boolean;
    silent: boolean;
}
interface TraceOptions extends App.TraceOptions {}
// interface InitOptions extends Options {}

const templateIssue = `${chalk.green('${uri}')}:${chalk.yellow('${row}:${col}')} - Unknown word (${chalk.red('${text}')})`;
const templateIssueLegacy = `${chalk.green('${uri}')}[\${row}, \${col}]: Unknown word: ${chalk.red('${text}')}`;
const templateIssueWordsOnly = '${text}';

function genIssueEmitter(template: string) {
    return function issueEmitter(issue: App.Issue) {
        console.log(formatIssue(template, issue));
    };
}

function errorEmitter(message: string, error: Error) {
    console.error(chalk.red(message), error);
    return Promise.resolve();
}

function infoEmitter(message: string, msgType: App.MessageType) {
    switch (msgType) {
        case 'Debug': console.info(chalk.cyan(message)); break;
        case 'Info': console.info(chalk.yellow(message)); break;
        case 'Progress': console.info(chalk.white(message)); break;
    }
}

function debugEmitter(message: string) {
    infoEmitter(message, App.MessageTypes.Debug);
}

function nullEmitter(_: string | App.Issue) {}
async function asyncNullEmitter(_: string | App.Issue) {}

function getEmitters(options: Options): App.Emitters {
    const issueTemplate = options.wordsOnly ? templateIssueWordsOnly : options.legacy ? templateIssueLegacy : templateIssue;
    const { silent = false, issues } = options;
    return {
        issue: silent || !issues ? nullEmitter : genIssueEmitter(issueTemplate),
        error: silent ? asyncNullEmitter : errorEmitter,
        info: silent || !options.verbose ? nullEmitter : infoEmitter,
        debug: options.debug ? debugEmitter : nullEmitter,
    };
}

export async function run(program?: commander.Command, argv?: string[]): Promise<void> {
    const prog = program || commander;
    const args = argv || process.argv;

    return new Promise((resolve, rejects) => {
        let showHelp = true;

        (prog as any).exitOverride();

        prog
            .version(npmPackage.version)
            .description('Spelling Checker for Code')
            .option(
                '-c, --config <cspell.json>',
                'Configuration file to use.  By default cspell looks for cspell.json in the current directory.'
            )
            .option('--no-color', 'Turn off color.')
            .option('--color', 'Force color');

        prog
            .option('-v, --verbose', 'display more information about the files being checked and the configuration')
            .option('--local <local>', 'Set language locals. i.e. "en,fr" for English and French, or "en-GB" for British English.')
            .option('--legacy', 'Legacy output')
            .option('--languageId <language>', 'Force programming language for unknown extensions. i.e. "php" or "scala"')
            .option('--wordsOnly', 'Only output the words not found in the dictionaries.')
            .option('-u, --unique', 'Only output the first instance of a word not found in the dictionaries.')
            .option('--debug', 'Output information useful for debugging cspell.json files.')
            .option('-e, --exclude <glob>', 'Exclude files matching the glob pattern')
            .option('--no-issues', 'Do not show the spelling errors.')
            .option('--no-summary', 'Turn off summary message in console')
            .option('-s, --silent', 'Silent mode, suppress error messages')
            .option('-r, --root <root folder>', 'Root directory, defaults to current directory.')
            // The following options are planned features
            // .option('-w, --watch', 'Watch for any changes to the matching files and report any errors')
            // .option('--force', 'Force the exit value to always be 0')
            .arguments('<files...>')
            .action((files: string[] | undefined, options: Options) => {
                const emitters: App.Emitters = getEmitters(options);
                if (!files || !files.length) {
                    return;
                }
                showHelp = false;
                return App.lint(files, options, emitters).then(
                    (result) => {
                        if (options.summary && !options.silent) {
                            console.error(
                                'CSpell: Files checked: %d, Issues found: %d in %d files',
                                result.files,
                                result.issues,
                                result.filesWithIssues.size
                            );
                        }
                        if (result.issues) {
                            throw new CheckFailed('check failed', 1);
                        }
                    }
                );
            });

        interface TraceCommandOptions {
            parent: TraceOptions;
        }
        prog
            .command('trace')
            .description('Trace words')
            .option(
                '-c, --config <cspell.json>',
                'Configuration file to use.  By default cspell looks for cspell.json in the current directory.'
            )
            .option('--local <local>', 'Set language locals. i.e. "en,fr" for English and French, or "en-GB" for British English.')
            .option('--languageId <language>', 'Force programming language for unknown extensions. i.e. "php" or "scala"')
            .option('--no-color', 'Turn off color.')
            .option('--color', 'Force color')
            .arguments('<words...>')
            .action((words: string[], options: TraceCommandOptions) => {
                showHelp = false;
                return App.trace(words, options.parent).then(
                    (result) => {
                        result.forEach(emitTraceResult);
                    }
                );
            });

        interface CheckCommandOptions {
            parent: BaseOptions;
        }

        prog
            .command('check <files...>')
            .description('Spell check file(s) and display the result. The full file is displayed in color.')
            .option(
                '-c, --config <cspell.json>',
                'Configuration file to use.  By default cspell looks for cspell.json in the current directory.'
            )
            .option('--no-color', 'Turn off color.')
            .option('--color', 'Force color')
            .action(async (files: string[], options: CheckCommandOptions) => {
                showHelp = false;

                for (const filename of files) {
                    console.log(chalk.yellowBright(`Check file: ${filename}`));
                    console.log();
                    try {
                        const result = await checkText(filename, options.parent);
                        for (const item of result.items) {
                            const fn =
                                item.flagIE === App.IncludeExcludeFlag.EXCLUDE ? chalk.gray : item.isError ? chalk.red : chalk.whiteBright;
                            const t = fn(item.text);
                            process.stdout.write(t);
                        }
                        console.log();
                    } catch (e) {
                        console.error(`Failed to read "${filename}"`);
                    }
                    console.log();
                }
            });

        /*
        program
            .command('init')
            .description('(Alpha) Initialize a cspell.json file.')
            .option('-o, --output <cspell.json>', 'define where to write file.')
            .option('--extends <cspell.json>', 'extend an existing cspell.json file.')
            .action((options: InitOptions) => {
                showHelp = false;
                CSpellApplication.createInit(options).then(
                    () => process.exit(0),
                    () => process.exit(1)
                );
                console.log('Init');
            });
        */

        const usage = `

Examples:
    cspell "*.js"                   Check all .js files in the current directory
    cspell "**/*.js"                Check all .js files from the current directory
    cspell "src/**/*.js"            Only check .js under src
    cspell "**/*.txt" "**/*.js"     Check both .js and .txt files.
    cat LICENSE | cspell stdin      Read from stdin the contents of LICENSE
`;

        prog.on('--help', function () {
            console.log(usage);
            showHelp = false;
        });

        function reject(e: any) {
            if (showHelp) {
                prog.help();
            }
            rejects(e);
        }

        try {
            prog.parseAsync(args).then(() => {
                if (showHelp) {
                    prog.help();
                }
                resolve();
            }).catch(reject);
        } catch (e) {
            reject(e);
        }
    });
}

function emitTraceResult(r: App.TraceResult) {
    const terminalWidth = process.stdout.columns || 120;
    const widthName = 20;
    const w = chalk.green(r.word);
    const f = r.found ? chalk.whiteBright('*') : chalk.dim('-');
    const n = chalk.yellowBright(pad(r.dictName, widthName));
    const used = [r.word.length, 1, widthName].reduce((a, b) => a + b, 3);
    const widthSrc = terminalWidth - used;
    const s = chalk.white(trimMid(r.dictSource, widthSrc));
    const line = [w, f, n, s].join(' ');
    console.log(line);
}

function pad(s: string, w: number): string {
    return (s + ' '.repeat(w)).substr(0, w);
}

function trimMid(s: string, w: number): string {
    if (s.length <= w) {
        return s;
    }
    const l = Math.floor((w - 3) / 2);
    const r = Math.ceil((w - 3) / 2);
    return s.substr(0, l) + '...' + s.substr(-r);
}

function formatIssue(template: string, issue: App.Issue) {
    const { uri = '', row, col, text } = issue;
    return template
        .replace(/\$\{uri\}/, uri)
        .replace(/\$\{row\}/, row.toString())
        .replace(/\$\{col\}/, col.toString())
        .replace(/\$\{text\}/, text);
}

export class CheckFailed extends Error {
    constructor(message: string, readonly errorCode: number, ) {
        super(message);
    }
}
