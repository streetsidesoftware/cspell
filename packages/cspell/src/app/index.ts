export * from './application.js';
export { getReporter as getDefaultReporter } from './cli-reporter.js';
export type { BaseOptions, LinterCliOptions as CSpellApplicationOptions, TraceOptions } from './options.js';
export * from '@cspell/cspell-types';
