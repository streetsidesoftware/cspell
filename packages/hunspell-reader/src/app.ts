// cSpell:ignore findup
import { program as commander } from 'commander';
import * as fs from 'fs-extra';
import type { Sequence } from 'gensequence';
import { genSequence } from 'gensequence';

import { asAffWord } from './aff';
import type { AffWord } from './affDef';
import { IterableHunspellReader } from './IterableHunspellReader';
import { iterableToStream } from './iterableToStream';
import { batch, uniqueFilter } from './util';

const uniqueHistorySize = 500000;

// eslint-disable-next-line @typescript-eslint/no-var-requires
const packageInfo = require('../package.json');
const version = packageInfo['version'];

let displayHelp = true;
let logStream: NodeJS.WritableStream = process.stderr;

commander.version(version);

commander
    .command('words <hunspell_dic_file>')
    .option('-o, --output <file>', 'output file - defaults to stdout')
    .option('-s, --sort', 'sort the list of words')
    .option('-u, --unique', 'make sure the words are unique.')
    .option('-l, --lower_case', 'output in lower case')
    .option('-T, --no-transform', 'Do not apply the prefix and suffix transforms.  Root words only.')
    .option('-x, --infix', 'Return words with prefix / suffix breaks. ex: "un<do>ing"')
    .option('-r, --rules', 'Append rules used to generate word.')
    .option('-p, --progress', 'Show progress.')
    .option('-m, --max_depth <limit>', 'Maximum depth to apply suffix rules.')
    .option('-n, --number <limit>', 'Limit the number of words to output.')
    .option('--forbidden', 'include forbidden words')
    .option('--partial_compounds', 'include words that must be part of a compound word')
    .option('--only_forbidden', 'includes only words that are forbidden')
    .description('Output all the words in the <hunspell.dic> file.')
    .action(action);

commander.parse(process.argv);

if (displayHelp) {
    commander.help();
}

function notify(message: string, newLine = true) {
    message = message + (newLine ? '\n' : '');
    logStream.write(message, 'utf-8');
}

function yesNo(value: boolean) {
    return value ? 'Yes' : 'No';
}

function affWordToInfix(aff: AffWord): AffWord {
    return { ...aff, word: aff.prefix + '<' + aff.base + '>' + aff.suffix };
}

function mapWord(map: (word: string) => string): (aff: AffWord) => AffWord {
    return (aff: AffWord) => ({ ...aff, word: map(aff.word) });
}

function appendRules(aff: AffWord): AffWord {
    return { ...aff, word: aff.word + '\t[' + aff.rulesApplied + ' ]\t' + '(' + aff.dic + ')' };
}

function writeSeqToFile(seq: Sequence<string>, outFile: string | undefined): Promise<void> {
    return new Promise((resolve, reject) => {
        let resolved = false;
        const out = outFile ? fs.createWriteStream(outFile) : process.stdout;
        const bufferedSeq = genSequence(batch(seq, 500)).map((batch) => batch.join(''));
        const dataStream = iterableToStream(bufferedSeq);
        const fileStream = dataStream.pipe(out);
        const endEvents = ['finish', 'close', 'end'];

        function resolvePromise() {
            if (!resolved) {
                resolved = true;
                resolve();
            }
        }
        const endHandler = () => {
            cleanupStreams();
            setTimeout(resolvePromise, 10);
        };
        const errorHandler = (e: Error) => {
            cleanupStreams();
            reject(e);
        };

        listenToStreams();

        function listenToStreams() {
            endEvents.forEach((event) => fileStream.addListener(event, endHandler));
            fileStream.addListener('error', errorHandler);
            dataStream.addListener('end', endHandler);
        }

        function cleanupStreams() {
            endEvents.forEach((event) => fileStream.removeListener(event, endHandler));
            fileStream.removeListener('error', errorHandler);
            dataStream.removeListener('end', endHandler);
        }
    });
}

function action(hunspellDicFilename: string, options: Options): Promise<void> {
    return actionPrime(hunspellDicFilename, options).catch((reason: { code: string }) => {
        if (reason.code === 'EPIPE') {
            console.log(reason);
            return;
        }
        console.error(reason);
    });
}

interface Options {
    sort?: boolean;
    unique?: boolean;
    lower_case?: boolean;
    output?: string;
    transform?: boolean;
    infix?: boolean;
    rules?: boolean; // append rules
    progress?: boolean; // show progress
    number?: string; // limit the number to output
    max_depth?: string; // limit the recursive depth to apply suffixes.
    forbidden: boolean; // include forbidden words
    only_forbidden: boolean; // only include forbidden words
    partial_compounds: boolean; // include partial word compounds
}

async function actionPrime(hunspellDicFilename: string, options: Options) {
    displayHelp = false;
    const {
        sort = false,
        unique = false,
        output: outputFile,
        lower_case: lowerCase = false,
        transform = true,
        infix = false,
        rules = false,
        progress: showProgress = false,
        max_depth,
        forbidden = false,
        only_forbidden: onlyForbidden = false,
        partial_compounds: partialCompoundsAllowed = false,
    } = options;
    logStream = outputFile ? process.stdout : process.stderr;
    const log = notify;
    log('Write words');
    log(`Sort: ${yesNo(sort)}`);
    log(`Unique: ${yesNo(unique)}`);
    const baseFile = hunspellDicFilename.replace(/\.(dic|aff)$/, '');
    const dicFile = baseFile + '.dic';
    const affFile = baseFile + '.aff';
    log(`Dic file: ${dicFile}`);
    log(`Aff file: ${affFile}`);
    log(`Generating Words...`);
    const reader = await IterableHunspellReader.createFromFiles(affFile, dicFile);
    if (max_depth && Number.parseInt(max_depth) >= 0) {
        reader.maxDepth = Number.parseInt(max_depth);
    }
    const transformers: ((aff: AffWord) => AffWord)[] = [];
    const filters: ((aff: AffWord) => boolean)[] = [];
    if (!forbidden && !onlyForbidden) filters.push((aff) => !aff.flags.isForbiddenWord);
    if (onlyForbidden) filters.push((aff) => !!aff.flags.isForbiddenWord);
    if (!partialCompoundsAllowed) filters.push((aff) => !aff.flags.isOnlyAllowedInCompound);
    if (infix) {
        transformers.push(affWordToInfix);
    }
    if (lowerCase) {
        transformers.push(mapWord((a) => a.toLowerCase()));
    }
    if (rules) {
        transformers.push(appendRules);
    }
    transformers.push(mapWord((a) => a.trim()));
    const dicSize = reader.dic.length;
    let current = 0;
    const calcProgress = () => '\r' + current + ' / ' + dicSize;
    const reportProgressRate = 253;
    const callback = showProgress
        ? () => {
              current++;
              !(current % reportProgressRate) && process.stderr.write(calcProgress(), 'utf-8');
          }
        : () => {
              /* void */
          };
    const seqWords = transform ? reader.seqAffWords(callback) : reader.seqRootWords().map(asAffWord);
    const filterUnique = unique ? uniqueFilter<string>(uniqueHistorySize) : (_: string) => true;

    const applyTransformers = (aff: AffWord) => transformers.reduce((aff, fn) => fn(aff), aff);
    const applyFilters = (aff: AffWord) => filters.reduce((cur, fn) => cur && fn(aff), true);

    const allWords = seqWords
        .filter(applyFilters)
        .map(applyTransformers)
        .map((a) => a.word)
        .filter((a) => !!a)
        .filter(filterUnique)
        .map((a) => a + '\n');

    const words = options.number ? allWords.take(Number.parseInt(options.number)) : allWords;

    if (sort) {
        log('Sorting...');
        const data = words.toArray().sort().join('');
        const fd = outputFile ? fs.openSync(outputFile, 'w') : 1;
        fs.writeSync(fd, data);
    } else {
        await writeSeqToFile(words, outputFile);
    }
    if (showProgress) {
        console.error(calcProgress());
    }
    log('Done.');
}
