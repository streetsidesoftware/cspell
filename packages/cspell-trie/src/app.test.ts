import * as Commander from 'commander';
import * as Path from 'path';
import type { Constructable } from 'vitest';
import { afterEach, describe, expect, test, vi } from 'vitest';

import * as app from './app.js';

const projectRoot = Path.join(__dirname, '..');

function argv(...args: string[]): string[] {
    return [...process.argv.slice(0, 2), ...args];
}

function getCommander() {
    return new Commander.Command();
}

function pathRoot(...parts: string[]): string {
    return Path.join(projectRoot, ...parts);
}

function pathSamples(...parts: string[]): string {
    return pathRoot('samples', ...parts);
}

function pathTemp(...parts: string[]): string {
    return pathRoot('temp', ...parts);
}

type ErrorResult = undefined | Constructable | string | RegExp;
type Test = [string, string[], ErrorResult];
const tests: Test[] = [
    t('No Args', [], Commander.CommanderError),
    t('create file', ['create', pathSamples('softwareTerms.txt'), '-o', pathTemp('samples.trie')], undefined),
    t('read file', ['reader', pathTemp('samples.trie'), '-o', pathTemp('test.txt')], undefined),
];

describe('Validate App', () => {
    afterEach(() => {
        vi.restoreAllMocks();
    });
    test.each(tests)('%s', async (_: string, args: string[], errorResult: ErrorResult) => {
        const out = createCollector();

        const commander = getCommander();
        commander.configureOutput({ writeOut: out.stdout, writeErr: out.stderr });
        commander.exitOverride();

        vi.spyOn(console, 'error').mockImplementation(out.error);
        vi.spyOn(console, 'log').mockImplementation(out.log);
        vi.spyOn(console, 'info').mockImplementation(out.info);
        vi.spyOn(console, 'warn').mockImplementation(out.warn);

        const result = app.run(commander, argv(...args));
        if (errorResult) {
            await expect(result).rejects.toThrow(errorResult);
        } else {
            await expect(result).resolves.not.toThrow();
        }
        expect(out.getText()).toMatchSnapshot();
    });
});

function t(testName: string, args: string[], errorResult: ErrorResult): Test {
    return [testName, args, errorResult];
}

interface Collector {
    log(text: string): void;
    info(text: string): void;
    warn(text: string): void;
    error(text: string): void;
    stdout(text: string): void;
    stderr(text: string): void;
    lines: string[];
    getText(): string;
}

function createCollector(): Collector {
    const lines: string[] = [];

    function handler(prefix: string): (text: string) => void {
        return function (text: string) {
            lines.push(prefix + text.split('\n').join('\n' + prefix));
        };
    }
    function getText(): string {
        return lines.join('\n');
    }
    const log = handler('[log]: ');
    const info = handler('[info]: ');
    const warn = handler('[warn]: ');
    const error = handler('[error]: ');
    const stdout = handler('[stdout]: ');
    const stderr = handler('[stderr]: ');

    return { lines, getText, log, info, warn, error, stdout, stderr };
}
