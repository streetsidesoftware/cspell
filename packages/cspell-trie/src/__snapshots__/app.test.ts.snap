// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Validate App > No Args 1`] = `
"[stderr]: Usage: worker [options] [command]
[stderr]: 
[stderr]: Options:
[stderr]:   -V, --version                 output the version number
[stderr]:   -h, --help                    display help for command
[stderr]: 
[stderr]: Commands:
[stderr]:   create [options] <file.txt>   Generate a file for use with cspell
[stderr]:   reader [options] <file.trie>  Read a cspell trie file and output the list of
[stderr]:                                 words.
[stderr]:   help [command]                display help for command
[stderr]: "
`;

exports[`Validate App > create file 1`] = `
"[log]: Create Trie
[log]: Generating...
[log]: Processing Trie
[log]: Export Trie"
`;

exports[`Validate App > read file 1`] = `"[log]: Reading Trie"`;
